Verify product dependencies w/ target platform, vs. hardcoded `macOS`

peterkos opened the pull request:
Product dependencies are now verified using the `buildEnvironment`'s platform, instead of the hardcoded `macOS` platform.

This builds on the work of #6732, now allowing cross-compilation of product dependencies 🎉

### Motivation:
Let's say we're cross-compiling macOS -> iOS simulator, with a host system of `macOS v13`.
```swift
let package = Package(
    name: "SwiftFlashlight",
    products: [.library(...)],
    platforms: [.iOS(.v16)],
    dependencies: [
        // internally, for iOS(.v16), macOS(.v14)
        .package(url: "probably/alamofire", exact: "1.2.3")
    ],
    // ...
)
```

As long as we set our triple & SDK, we _should_ be good... (and, with #6828, only the triple needed)
```fish
> swift build --triple arm64-apple-ios-simulator \
              --sdk "$(xcrun --sdk iphonesimulator --show-sdk-path)"
...
error: the library 'SwiftFlashlight' requires macos 10.13,
but depends on the product 'probably/alamofire' which requires macos 10.14;  [...]
```

Even though `SwiftFlashlight` only supports `iOS(.v16)`, [SupportedPlatform](https://github.com/apple/swift-package-manager/blob/main/Documentation/PackageDescription.md#supportedplatform) "[by] default, [...] assigns a predefined minimum deployment version for each supported platforms unless you configure supported platforms [...]".  As a side-effect, `macOS` is inferred as a valid platform. (This is why no "target/platform mismatch" error was thrown.)

Previously, product dependencies were verified against `macOS`, rather than the target platform. Since `macOS` is always inferred, this surfaced as a "woah! your dependencies are out of sync".

```swift

let swiftFlashlight = Package(
    platforms: [
        .iOS(.v16),
        .macOS(.v13) // <-- inferred-|
    ],                               |
)                                    |-- // since --triple is .iOS,
                                     |   // shouldn't be comparing
let probably/alamofire = Package (   |   // macOS.
    platforms: [                     |
        iOS(.v16),                   |
        macOS(.v14)] // <-- inferred-|
    ]
)


```


Now, the intention of `--triple` is respected 🐱

### Modifications:
- `validateDeploymentVersionOfProductDependency` accepts a `buildEnvironment` containing the target platform

### Result:

Product dependencies can now be compiled for `darwin` other than macOS -- i.e, iOS :)

### Future work
- Target-level or product-level platform support is still up in the air
- This only applies to `triple.isDarwin()` -- perhaps a next step of, verifying dependencies against non-darwin?

peterkos commented:
This was a fun one to chase down 😅 
(note to self: `Apple Swift version 5.9 (swift-5.9-RELEASE)`)

Some questions:
- I tried running `swiftformat` but it changed 600+ files -- just `BuildPlan` was ~300 lines. Thoughts?
- Happy to add some tests around this; wanted to get some thoughts before diving in there :)

MaxDesiatov commented:
> I tried running swiftformat but it changed 600+ files -- just BuildPlan was ~300 lines 😵, thoughts?

Please run `Utilities/soundness.sh` script, which will only format files that were changed.

> Happy to add some tests around this; wanted to get some thoughts before diving in there :)

Yes, please add the tests, that would be great. Thanks!

peterkos commented:
Thanks! I'll take a crack at it soon :)

peterkos commented:
@MaxDesiatov Apologies for the delay on this! One test added, one test modified 🧪 

Since `utilities/soundness.sh` works from the `git diff`, I added a newline in each file I modified + ran the script as a separate commit for readability. As you can see, those files haven't had a run in a _while_, heh.

MaxDesiatov commented:
Thanks! Just as heads up, this will conflict with https://github.com/apple/swift-package-manager/pull/7161

peterkos commented:
Resolved conflicts 👍 

Looks like `soundness.sh` disagrees with how Swift requires raw string literals to be formatted tho. One example:
```swift
// BuildPlanTests L#5749
func testArtifactsArchiveBinaryTargets(...) {
    // ...
                        "path": "all-platforms/mytool",
                        "supportedTriples": ["\(
                            artifactTriples.map(\.tripleString)
                                .joined(separator: "\", \"")
    )"]
                }
            ]
        }
    }
}
                    """
                //   ^ swiftformat throws this here,
                //     but is a swift build error :(
```

I've manually reverted the script's raw string changes for now, but it seems like the way swiftformat expands inline closures (among other things) could be improved -- that's my guess 🚀 

Also the `soundness.sh` script didn't print any output unless I tweaked L60. Assuming it's a me-issue, but can push up that change if useful.
```sh
swiftformat $changed_file > dev/null 2>&1 # old
swiftformat $changed_file 2>&1            # new
```

MaxDesiatov commented:
@swift-ci test

MaxDesiatov commented:
@swift-ci test windows

MaxDesiatov commented on Sources/SPMTestSupport/MockBuildTestHelper.swift:
```diff
@@ -134,23 +133,29 @@ enum BuildError: Swift.Error {
         self.plan = plan
-        self.productMap = try Dictionary(throwingUniqueKeysWithValues: plan.buildProducts.compactMap { $0 as? Build.ProductBuildDescription }.map{ ($0.product.name, $0) })
-        self.targetMap = try Dictionary(throwingUniqueKeysWithValues: plan.targetMap.map{ ($0.0.name, $0.1) })
+        self
+            .productMap = try Dictionary(
+                throwingUniqueKeysWithValues: plan.buildProducts
+                    .compactMap { $0 as? Build.ProductBuildDescription }.map { (
+                        $0.product.name,
+                        $0
+                    ) }
+            )
```
```suggestion
        self.productMap = try Dictionary(
                throwingUniqueKeysWithValues: plan.buildProducts
                    .compactMap { $0 as? Build.ProductBuildDescription }
                    .map { ($0.product.name, $0) }
        )
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -1057,9 +1143,22 @@ final class BuildPlanTests: XCTestCase {
+        XCTAssertMatch(
+            exe,
+            [
+                "-swift-version",
+                "4",
```
```suggestion
                "-swift-version", "4",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -1057,9 +1143,22 @@ final class BuildPlanTests: XCTestCase {
+                "-O",
+                .equal(self.j),
+                "-DSWIFT_PACKAGE",
+                "-module-cache-path",
+                "\(buildPath.appending(components: "ModuleCache"))",
```
```suggestion
                "-module-cache-path", "\(buildPath.appending(components: "ModuleCache"))",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -1134,9 +1235,22 @@ final class BuildPlanTests: XCTestCase {
+        XCTAssertMatch(
+            exe,
+            [
+                "-swift-version",
+                "4",
```
```suggestion
                "-swift-version", "4",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -1134,9 +1235,22 @@ final class BuildPlanTests: XCTestCase {
+                "-O",
+                .equal(self.j),
+                "-DSWIFT_PACKAGE",
+                "-module-cache-path",
+                "\(buildPath.appending(components: "ModuleCache"))",
```
```suggestion
                "-module-cache-path", "\(buildPath.appending(components: "ModuleCache"))",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -1831,39 +1980,46 @@ final class BuildPlanTests: XCTestCase {
+                "-Xlinker", "-bundle",
+                "-Xlinker", "-rpath", "-Xlinker", "@loader_path/../../../",
+                "@\(buildPath.appending(components: "PkgPackageTests.product", "Objects.LinkFileList"))",
+            ] +
+                rpathsForBackdeployment +
+                ["-target", "\(hostTriple.tripleString(forPlatformVersion: version))",
```
```suggestion
            ] + rpathsForBackdeployment + [
                "-target", "\(hostTriple.tripleString(forPlatformVersion: version))",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -1831,39 +1980,46 @@ final class BuildPlanTests: XCTestCase {
+                 "-Xlinker", "-add_ast_path", "-Xlinker",
+                 buildPath.appending(components: "Modules", "Foo.swiftmodule").pathString,
+                 "-Xlinker", "-add_ast_path", "-Xlinker",
+                 buildPath.appending(components: "Modules", "FooTests.swiftmodule").pathString,
+                 "-g"]
```
```suggestion
                 buildPath.appending(components: "Modules", "FooTests.swiftmodule").pathString,
                 "-g"
             ]
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -1926,9 +2084,24 @@ final class BuildPlanTests: XCTestCase {
+            exe,
+            [
+                .anySequence,
+                "-swift-version",
+                "4",
```
```suggestion
                "-swift-version", "4",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -1926,9 +2084,24 @@ final class BuildPlanTests: XCTestCase {
+                "-O",
+                .equal(self.j),
+                "-DSWIFT_PACKAGE",
+                "-module-cache-path",
+                "\(buildPath.appending(components: "ModuleCache"))",
```
```suggestion
                "-module-cache-path", "\(buildPath.appending(components: "ModuleCache"))",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -2267,9 +2447,28 @@ final class BuildPlanTests: XCTestCase {
+        XCTAssertMatch(
+            try result.target(for: "exe").swiftTarget().compileArguments(),
+            [
+                "-swift-version",
+                "4",
```
```suggestion
                "-swift-version", "4",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -2267,9 +2447,28 @@ final class BuildPlanTests: XCTestCase {
+                .equal(self.j),
+                "-DSWIFT_PACKAGE",
+                "-DDEBUG",
+                "-Xcc",
+                "-fmodule-map-file=\(Clibgit.appending(components: "module.modulemap"))",
+                "-module-cache-path",
+                "\(buildPath.appending(components: "ModuleCache"))",
```
```suggestion
                "-Xcc", "-fmodule-map-file=\(Clibgit.appending(components: "module.modulemap"))",
                "-module-cache-path", "\(buildPath.appending(components: "ModuleCache"))",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -2280,10 +2479,11 @@ final class BuildPlanTests: XCTestCase {
             "-Xlinker", "-rpath", "-Xlinker", "/fake/path/lib/swift-5.5/macosx",
             "-target", defaultTargetTriple,
-            "-Xlinker", "-add_ast_path", "-Xlinker", buildPath.appending(components: "exe.build", "exe.swiftmodule").pathString,
+            "-Xlinker", "-add_ast_path", "-Xlinker",
+            buildPath.appending(components: "exe.build", "exe.swiftmodule").pathString,
```
```suggestion
            "-Xlinker", "-add_ast_path",
            "-Xlinker", buildPath.appending(components: "exe.build", "exe.swiftmodule").pathString,
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -2421,7 +2626,8 @@ final class BuildPlanTests: XCTestCase {
             "-Xlinker", "-rpath", "-Xlinker", "/fake/path/lib/swift-5.5/macosx",
             "-target", defaultTargetTriple,
-            "-Xlinker", "-add_ast_path", "-Xlinker", buildPath.appending(components: "Foo.build", "Foo.swiftmodule").pathString,
+            "-Xlinker", "-add_ast_path", "-Xlinker",
+            buildPath.appending(components: "Foo.build", "Foo.swiftmodule").pathString,
```
```suggestion
            "-Xlinker", "-add_ast_path",
            "-Xlinker", buildPath.appending(components: "Foo.build", "Foo.swiftmodule").pathString,
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -2436,10 +2642,11 @@ final class BuildPlanTests: XCTestCase {
             "-Xlinker", "-rpath", "-Xlinker", "/fake/path/lib/swift-5.5/macosx",
             "-target", defaultTargetTriple,
-            "-Xlinker", "-add_ast_path", "-Xlinker", buildPath.appending(components: "Modules", "Bar.swiftmodule").pathString,
+            "-Xlinker", "-add_ast_path", "-Xlinker",
+            buildPath.appending(components: "Modules", "Bar.swiftmodule").pathString,
```
```suggestion
            "-Xlinker", "-add_ast_path",
            "-Xlinker", buildPath.appending(components: "Modules", "Bar.swiftmodule").pathString,
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -2538,49 +2747,84 @@ final class BuildPlanTests: XCTestCase {
+        XCTAssertMatch(
+            exe,
+            [
+                "-swift-version",
+                "4",
```
```suggestion
                "-swift-version", "4",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -2538,49 +2747,84 @@ final class BuildPlanTests: XCTestCase {
+                .equal(self.j),
+                "-DSWIFT_PACKAGE",
+                "-DDEBUG",
+                "-module-cache-path",
+                "\(buildPath.appending(components: "ModuleCache"))",
```
```suggestion
                "-module-cache-path", "\(buildPath.appending(components: "ModuleCache"))",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -2538,49 +2747,84 @@ final class BuildPlanTests: XCTestCase {
+        XCTAssertMatch(
+            lib,
+            [
+                "-swift-version",
+                "4",
```
```suggestion
                "-swift-version", "4",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -2538,49 +2747,84 @@ final class BuildPlanTests: XCTestCase {
+                .equal(self.j),
+                "-DSWIFT_PACKAGE",
+                "-DDEBUG",
+                "-module-cache-path",
+                "\(buildPath.appending(components: "ModuleCache"))",
```
```suggestion
                "-module-cache-path", "\(buildPath.appending(components: "ModuleCache"))",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -2538,49 +2747,84 @@ final class BuildPlanTests: XCTestCase {
+            "-emit-library",
+            "@\(buildPath.appending(components: "lib.product", "Objects.LinkFileList"))",
+            "-target", defaultTargetTriple,
+            "-g", "-use-ld=lld", "-Xlinker", "-debug:dwarf",
```
```suggestion
            "-g", "-use-ld=lld",
            "-Xlinker", "-debug:dwarf",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -3116,8 +3374,9 @@ final class BuildPlanTests: XCTestCase {
         let lib = try result.target(for: "lib").clangTarget()
         let args = [
             "-target", "x86_64-unknown-windows-msvc", "-O0",
```
```suggestion
            "-target", "x86_64-unknown-windows-msvc",
            "-O0",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -3116,8 +3374,9 @@ final class BuildPlanTests: XCTestCase {
             "-target", "x86_64-unknown-windows-msvc", "-O0",
-            "-DSWIFT_PACKAGE=1", "-DDEBUG=1", "-fblocks", "-I", Pkg.appending(components: "Sources", "lib", "include").pathString,
-            "-gdwarf"
+            "-DSWIFT_PACKAGE=1", "-DDEBUG=1", "-fblocks", "-I",
+            Pkg.appending(components: "Sources", "lib", "include").pathString,
```
```suggestion
            "-DSWIFT_PACKAGE=1",
            "-DDEBUG=1",
            "-fblocks",
            "-I", Pkg.appending(components: "Sources", "lib", "include").pathString,
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -3129,14 +3388,14 @@ final class BuildPlanTests: XCTestCase {
             "-Xcc", "-I", "-Xcc", "\(Pkg.appending(components: "Sources", "lib", "include"))",
             "-module-cache-path", "\(buildPath.appending(components: "ModuleCache"))",
             .anySequence,
             "-g", "-use-ld=lld", "-Xcc", "-gdwarf",
```
```suggestion
            "-g",
            "-use-ld=lld",
            "-Xcc", "-gdwarf",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -3145,8 +3404,8 @@ final class BuildPlanTests: XCTestCase {
-             "-target", "x86_64-unknown-windows-msvc",
-             "-g", "-use-ld=lld", "-Xlinker", "-debug:dwarf",
+            "-target", "x86_64-unknown-windows-msvc",
+            "-g", "-use-ld=lld", "-Xlinker", "-debug:dwarf",
```
```suggestion
            "-g",
            "-use-ld=lld",
            "-Xlinker", "-debug:dwarf",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -4373,16 +4802,40 @@ final class BuildPlanTests: XCTestCase {
+            [
+                .anySequence,
+                "-emit-objc-header",
+                "-emit-objc-header-path",
+                "/path/to/build/debug/Foo.build/Foo-Swift.h",
```
```suggestion
                "-emit-objc-header-path", "/path/to/build/debug/Foo.build/Foo-Swift.h",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -4373,16 +4802,40 @@ final class BuildPlanTests: XCTestCase {
+            [
+                .anySequence,
+                "-emit-objc-header",
+                "-emit-objc-header-path",
+                "/path/to/build/debug/Foo.build/Foo-Swift.h",
```
```suggestion
                "-emit-objc-header-path", "/path/to/build/debug/Foo.build/Foo-Swift.h",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -5134,7 +5622,15 @@ final class BuildPlanTests: XCTestCase {
+            clibraryBasicArguments,
+            [
+                .anySequence,
+                "-I",
+                "\(Pkg.appending(components: "StaticLibrary.xcframework", "\(platform)-\(arch)", "Headers"))",
```
```suggestion
                "-I", "\(Pkg.appending(components: "StaticLibrary.xcframework", "\(platform)-\(arch)", "Headers"))",
```

MaxDesiatov commented on Tests/BuildTests/BuildPlanTests.swift:
```diff
@@ -5520,11 +6038,13 @@ final class BuildPlanTests: XCTestCase {
             "-target", defaultTargetTriple,
-            "-Xlinker", "-add_ast_path", "-Xlinker", buildPath.appending(components: "Modules", "lib.swiftmodule").pathString,
-            "-Xlinker", "-add_ast_path", "-Xlinker", buildPath.appending(components: "exe.build", "exe.swiftmodule").pathString,
+            "-Xlinker", "-add_ast_path", "-Xlinker",
+            buildPath.appending(components: "Modules", "lib.swiftmodule").pathString,
+            "-Xlinker", "-add_ast_path", "-Xlinker",
+            buildPath.appending(components: "exe.build", "exe.swiftmodule").pathString,
```
```suggestion
            "-Xlinker", "-add_ast_path",
            "-Xlinker", buildPath.appending(components: "Modules", "lib.swiftmodule").pathString,
            "-Xlinker", "-add_ast_path",
            "-Xlinker", buildPath.appending(components: "exe.build", "exe.swiftmodule").pathString,
```

MaxDesiatov commented:
@swift-ci test

MaxDesiatov reviewed (approved):
Thank you!

MaxDesiatov commented:
@swift-ci test windows
